app/uk/gov/hmrc/gform/gform/processor/FormProcessor.scala

```diff
@@ -405,6 +417,109 @@ class FormProcessor(
+          }
+          TaskIdTaskStatusMapping(taskIdStatus)
+        }
+      case _ => TaskIdTaskStatusMapping(Map.empty).pure[Future]
+    }
+
+  }
+
+  private def evalStatusLookup(
+    coordinates: NonEmptyList[Coordinates],
+    cache: CacheData,
+    envelope: EnvelopeWithMapping,
+    formModelOptics: FormModelOptics[DataOrigin.Browser]
+  )(implicit
+    hc: HeaderCarrier,
+    messages: Messages,
+    l: LangADT,
+    sse: SmartStringEvaluator
+  ) = {
+    val formModel = formModelOptics.formModelVisibilityOptics.formModel
+    val cannotStartYetResolver = CannotStartYetResolver.create(formModelOptics.formModelRenderPageOptics.formModel)
+    val notRequiredResolver = NotRequiredResolver.create(formModelOptics.formModelVisibilityOptics)
+    for {
+      statusesLookup <- coordinates
+                          .traverse { coordinate =>
+                            val dataForCoordinate: Set[VariadicValue] =
+                              formModelOptics.formModelVisibilityOptics.data
+                                .forCoordinate(coordinate)
+                            val hasTerminationPage = formModel.taskList
+                              .availablePages(coordinate)
+                              .exists(_.isTerminationPage)
+
+                            for {
+                              formHandlerResult <-
+                                validationService.validateFormModel(
+                                  cache,
+                                  envelope,
+                                  formModelOptics.formModelVisibilityOptics,
+                                  Some(coordinate)
+                                )
+                              validatedATLs =
+                                validationService.validateATLs(
+                                  formModel.taskList.availablePages(
+                                    coordinate
+                                  ),
+                                  formModelOptics.formModelVisibilityOptics
+                                )
+                            } yield {
+                              val taskStatus =
+                                if (dataForCoordinate.isEmpty) {
+                                  TaskStatus.NotStarted
+                                } else if (
+                                  formHandlerResult.isFormValid && !hasTerminationPage && validatedATLs.isValid
+                                ) {
+                                  TaskStatus.Completed
+                                } else {
+                                  TaskStatus.InProgress
+                                }
+                              coordinate -> taskStatus
+                            }
+                          }
+                          .map(notRequiredResolver.resolveNotRequired)
+                          .map(
+                            cannotStartYetResolver.resolveCannotStartYet
+                          )
```

Comment:
There seems to be same logic in `private def statuses` of `TaskListRenderingService.scala`.
If I understand this correctly `TaskListRenderingService.scala` should only use statuses precomputed here, right?

Reply:
At least we should have only one implementation of how task status is computed.

Reply:
`TaskListRenderingService` computes the statuses when visiting landing page. But we need to compute the status of task in FormProcessor to get `In progress` status in task. I think it would be more logical to calculate and update the status of only the selected task instead of computing the status of all tasks in FormProcessor.

Reply:
> I think it would be more logical to calculate and update the status of only the selected task instead of computing the status of all tasks in FormProcessor.

Not evaluating all task's statuses can lead to cache of `taskIdTaskStatus` not being up to date at all time, if there is task dependency like this for example
```
{
  "title": "Check eligibility",
  "id": "applicants",
  "sections": [
    {
      "title": "Applicant",
      "fields": [
        {
          "id": "name",
          "type": "text",
          "format": "text",
          "label": "name"
        },
        {
          "id": "age",
          "includeIf": "${taskStatus(aboutYouTask) = 'Completed'}",
          "type": "text",
          "format": "text",
          "label": "Age"
        }
      ]
    }
  ]
}
```

At this point we should perhaps ask if such a case should be even allowed. Maybe we should limit where `taskStatus` function can be used to avoid these situations.

Reply:
I see it's correct. It will calculate status only for tasks with taskId.